Add new no-children-prop rule (Fixes #720)

benstepp opened the pull request:
Prevents children being passed as props. Children should be actual children.

I've never written an eslint rule, so any suggestions are definitely appreciated.

ljharb commented:
LGTM

lencioni commented on docs/rules/no-children-prop.md:
````diff
@@ -0,0 +1,21 @@
+
+The following patterns are considered warnings:
+
+```js
````
These should use jsx, not js, for better syntax highlighting.

ljharb replied:
I've been told `js` gives better highlighting than `jsx`, even for JSX content.

lencioni replied:
Oh? I usually see lots of blocks with red backgrounds when using `js` on JSX code. Here's an example.

Before: https://github.com/Khan/aphrodite/tree/436325f8848db974e6f9ffba18afa6af1c426095#api
After: https://github.com/Khan/aphrodite/tree/a1fdb520fd9fcc50148a30de2c49c7c1253a604e#api

ljharb replied:
That does look better :-) either way's fine, but i'm pretty sure it's come up before. Maybe Github improved the highlighting since.

lencioni commented on index.js:
```diff
@@ -50,7 +50,8 @@ var rules = {
   'require-optimization': require('./lib/rules/require-optimization'),
-  'no-find-dom-node': require('./lib/rules/no-find-dom-node')
+  'no-find-dom-node': require('./lib/rules/no-find-dom-node'),
+  'no-children-prop': require('./lib/rules/no-children-prop')
```
I think this should go in the recommended config as well. Thoughts @yannickcr?

ljharb replied:
I agree, but that change could be considered breaking whereas this change is minor. If this is going to make it into v6 then that's fine, though.

lencioni replied:
I think this is ready for v6 so might as well include it in recommended.

lencioni commented on tests/lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,124 @@
+      code: 'React.createElement("div", {children: [React.createElement("div"), React.createElement("div")]});',
+      errors: [{message: 'Do not pass children as props.'}],
+      parserOptions: parserOptions
+    }
```
This rule should also warn when Components receive a `children` prop. For example, this is bad too:

```jsx
<MyComponent children='Children' />
```

lencioni commented on lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,71 @@
+        }
+      },
+      CallExpression: function(node) {
+        if (isCreateElementWithProps(node)) {
```
nit: I think it would be slightly nicer to use an early return here to keep your code flatter.

lencioni commented:
@benstepp You've done a really great job with your first rule! 🎊

lencioni commented on lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,71 @@
+        if (isTagName(node.parent.name.name) && node.name.name === 'children') {
+          context.report({
+            node: node,
+            message: 'Do not pass children as props.'
```
It might be nice to additionally tell people what they should do. How about this error message?

> Do not pass children as a prop. Instead, nest children between the opening and closing tags.

lencioni commented on lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,71 @@
+          if (childrenProp) {
+            context.report({
+              node: node,
+              message: 'Do not pass children as props.'
```
Same here. What do you think about this?

> Do not pass children as a prop. Instead, pass them as additional arguments to `React.createElement`.

benstepp commented:
I went ahead and made changes to the error messages, and added the additional clarity regarding the desired behavior to the introduction of the docs.

* Added a few tests for a `<MyComponent />`.
* Switched to an early return to reduce the nesting.

lencioni commented on lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,65 @@
+        if (childrenProp) {
+          context.report({
+            node: node,
+            message: 'Do not pass children as props. Instead, pass them as the third argument to React.createElement.'
```
I believe that `React.createElement` can take an arbitrary number of arguments. Starting with the third argument they are considered children. This error message should probably be updated to reflect this.

lencioni commented on tests/lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,142 @@
+  invalid: [
+    {
+      code: '<div children="Children" />;',
+      errors: [{message: 'Do not pass children as props.'}],
```
I think you need to update your tests to match the new error messages.

lencioni commented on tests/lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,142 @@
+      parserOptions: parserOptions
+    },
+    {
+      code: '<MyComponent>Children</MyComponent>',
```
Will you please include a valid case using React.createElement with MyComponent and children?

lencioni commented on tests/lib/rules/no-children-prop.js:
```diff
@@ -0,0 +1,142 @@
+      parserOptions: parserOptions
+    },
+    {
+      code: 'React.createElement("div", {}, "Children");',
```
It might be nice to include an example with multiple children. e.g.

```js
React.createElement("div", {}, "Child 1", "Child 2");
```

lencioni replied:
Also, a valid case where there is at least one non-children prop with React.createElement and JSX.

benstepp commented:
Went ahead and added test cases with non-children props and multiple children, as well as clarified the error regarding `React.createElement`.

lencioni commented:
LGTM

lencioni commented:
@benstepp It looks like this needs a rebase.

lencioni commented:
@benstepp any chance you can rebase this?

benstepp commented:
Forgot I still had this over here.

Rebased over current master. Same 13 failing tests as master that would probably be better fixed as a separate commit.

lencioni commented:
Agreed on all accounts! Thanks for rebasing. 💃

lencioni commented on index.js:
```diff
@@ -96,6 +97,7 @@ module.exports = {
         'react/jsx-no-undef': 2,
         'react/jsx-uses-react': 2,
         'react/jsx-uses-vars': 2,
+        'react/no-danger': 2,
```
I think you unintentionally added this line, which is causing specs to fail.

benstepp replied:
Looks like you're right. Went ahead and removed that line.

lencioni commented:
Thanks!
